### PR TITLE
fix: flash modals stack up and are not visible until scrolled to top

### LINKED ISSUE #35405: [Beta] Setting page update modal should be removed
## Describe the bug
On updating preferences on the settings page. There's a modal which shows up on the top of the page which I think is of no use as they are not visible until we scroll to the top. If we change multiple preferences then the modals stack up. 

## Screenshots
screenshot 23

## Solution proposal
We should show a side modal and then it should get closed automatically after a few seconds. Something like:
screenshot 24

## Desktop (please complete the following information):
 - OS: Windows 10 Pro 32 bit
 - Browser: Chrome
 - Version: Latest stable

### PR DESCRIPTION
**Pull request by hemakshis**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #35405

### CODE REVIEW

**Comment by hemakshis** [comment]
Hi! The animation part is yet to be added. I looked for how we can make transitions in React.js but I guess we want to add a transition whenever the Flash component mounts or unmounts. So this means we may have to change the component to a Class component or else there's a library [react-transition-group](https://github.com/reactjs/react-transition-group) so just wanted some opinion on this.

**Review by ojeytonwilliams — COMMENTED** [review]
I like what you've done so far - it's already looking much better!

As for the animations, I think that `react-transition-group` is the way to go.  That should make it easy to animate the flash messages being replaced or dismissed.  `react-transition-group` is already used by `Alert` internally, but I suspect that you will end up needing to wrap the flash messages in order to animate them being replaced. 

Also, could you make sure it's not necessary to scroll up to see the message?

**Comment by hemakshis** [comment]
Hi @ojeytonwilliams, thanks for the reviews :)
I saw that we have a `Fade` component in our `@freecodecamp/react-bootstrap` library which internally uses the `react-transition-group` library.

It takes a prop `in` (`boolean`) which is responsible for triggering a "fade" effect. Now, since we render the `Flash` component on the value of `hasMessage`, but whenever the user makes a 2nd (or more) change(s) to any other settings a `createFlashMessage` action automatically gets triggered but the value of `hasMessage` remains to be true (because message is already there it's just that the value has changed) and that is why the `Fade` component cannot re-trigger the effect (as the value of `in` prop remains unchanged)

I tried all the possible ways I could think of but can't make it work. Really appreciate some help in here.

**Comment by ojeytonwilliams** [comment]
@hemakshis Sorry for the delay getting back to you - this was a bit more complicated than I realised!  You're right that there is a `Fade` component in that library, but unfortunately `Alert` does not use it (not in the version of the library we use).  

My advice is to ignore `Alert` and instead wrap the `Flash` component in a `CSSTransition` and a `TransitionGroup`.  They will handle the animation of the child component.  Take a look at [this example](https://codesandbox.io/s/00rqyo26kn?from-embed) and the [docs](https://reactcommunity.org/react-transition-group/).  Also, good point about `hasMessage` - that isn't going to change between messages.  You can use `flashMessage.id`, since that is intended to be a unique identifier of the message.

Feel free to send me a message on Gitter if any of that was unclear.

**Comment by hemakshis** [comment]
Hi @ojeytonwilliams, thanks for the help! It worked but I feel the animation results are not so satisfactory :/
If once you can also have a look and tell what you feel would be great!

**Comment by ojeytonwilliams** [comment]
@hemakshis You're very welcome!  The main thing you're missing is the animation css.  If you take a look [here](https://reactcommunity.org/react-transition-group/css-transition) you can see the kinds of rules you need to specify.  In order to apply the right classes, you also have to pass `CSSTransition` the `classNames` (not `className`) prop.  Have a play with those.

Finally, if you want to animate the message appearing and being dismissed (rather than just being replaced), you need to wrap

```jsx
{hasMessage ? (
  <Flash
       flashMessage={flashMessage}
       onClose={removeFlashMessage}
    />
) : null}
```
with a `TransitionGroup`, otherwise the whole thing leaves the DOM and nothing gets animated.  I'm not sure if that's what we want, but if it is - that's how you do it!

**Comment by hemakshis** [comment]
> Finally, if you want to animate the message appearing and being dismissed (rather than just being replaced), you need to wrap
> 
> ```js-jsx
> {hasMessage ? (
>   <Flash
>        flashMessage={flashMessage}
>        onClose={removeFlashMessage}
>     />
> ) : null}
> ```
> 
> with a `TransitionGroup`, otherwise the whole thing leaves the DOM and nothing gets animated. I'm not sure if that's what we want, but if it is - that's how you do it!

Instead of wrapping the Flash component in `DefaulLayout` component, I have done that inside the Flash component itself, like this:-
```javascript
function Flash({ flashMessage, onClose }) {
  const { type, message, id } = flashMessage;
  return (
    <TransitionGroup>
      <CSSTransition key={id} timeout={500} classNames="flash-message">
        <Alert
          bsStyle={type}
          className='flash-message'
          key={id}
          onDismiss={createDismissHandler(onClose, id)}
        >
          <div dangerouslySetInnerHTML={{ __html: message }} />
        </Alert>
      </CSSTransition>
    </TransitionGroup>
  );
}
```
I have added the CSS Animations part and kept the Flash Messages fixed to the top. Just please have a look.

File: client/src/components/Flash/index.js [hunk 1]
```diff
@@ -1,35 +1,39 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import { Alert } from '@freecodecamp/react-bootstrap';
+import { TransitionGroup, CSSTransition } from 'react-transition-group';
 
 import './flash.css';
 
 function createDismissHandler(fn, id) {
```
Comment (ojeytonwilliams) [inline client/src/components/Flash/index.js]:
```suggestion
```

File: client/src/components/Flash/index.js [hunk 2]
```diff
@@ -1,35 +1,39 @@
...
   return () => fn(id);
```
Comment (ojeytonwilliams) [inline client/src/components/Flash/index.js]:
```suggestion
```

File: client/src/components/Flash/index.js [hunk 3]
```diff
@@ -1,35 +1,39 @@
...
 }
```
Comment (ojeytonwilliams) [inline client/src/components/Flash/index.js]:
```suggestion
```

File: client/src/components/Flash/index.js [hunk 4]
```diff
@@ -1,35 +1,39 @@
...
-  ));
+function Flash({ flashMessage, onClose }) {
+  const { type, message, id } = flashMessage;
+  return (
+    <TransitionGroup>
+      <CSSTransition classNames='flash-message' key={id} timeout={500}>
+        <Alert
+          bsStyle={type}
+          className='flash-message'
+          key={id}
```
Comment (ojeytonwilliams) [inline client/src/components/Flash/index.js]:
```suggestion
```

File: client/src/components/Flash/index.js [hunk 5]
```diff
@@ -1,35 +1,39 @@
...
+          onDismiss={createDismissHandler(onClose, id)}
```
Comment (ojeytonwilliams) [inline client/src/components/Flash/index.js]:
```suggestion
          onDismiss={onClose}
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
It's looking great!

The changes I've requested are just to clean things up slightly.  Since we no longer support multiple messages, the code can be simplified a bit.

**Review by ojeytonwilliams — APPROVED** [review]

**Comment by ojeytonwilliams** [comment]
Thanks for sorting that out @hemakshis, much appreciated!

Everything LGTM, but could you make sure it's okay, @Bouncey?

**Comment by scissorsneedfoodtoo** [comment]
Can you take a look at this @ahmadabdolsaheb?

**Comment by moT01** [comment]
I checked this out locally, functionally it looks like it works. Overall it looks pretty good. The flash messages could possible use some darker colors when in night mode. When the message pops up and then you change another setting to get the same message to pop up - all it does is fade out and in a tiny bit. I would like to see something more obvious possibly. A solution I like for this would possibly be to have the messages say exactly what you changed. This would probably have to be another issue and PR. But instead of it saying "We have updated your preferences" and then fading out and in slightly to only say "We have updated your preferences" again. It would say something like "We have changed your heat map to private" and then the same message would never pop up twice if it said something specific for each item - making it obvious what setting was changed.

As far as the code goes - it looks good to me, but we probably want someone more familiar with how we want the code to be to take a look.

Nice job @hemakshis!

On another note - pretty sure it has nothing to do with this PR, but when in night mode - changing certain settings will turn off night mode

**Comment by ValeraS** [comment]
Is it good that the error message would be replaced with an info one?

**Comment by raisedadead** [comment]
> Is it good that the error message would be replaced with an info one?

Yes, we want to reserve the danger (red) error messages for extremely critical failures. Error messages are distracting, and we should use them to convey urgency only.

Everything like known minor issues are conveyed as info and success messages. Even success messages are distraction IMHO.

**Comment by ValeraS** [comment]
I mean, if the error message is shown it will be replaced a next info message, is it right? Or should we display both simultaneously?

**Comment by raisedadead** [comment]
> I mean, if the error message is shown it will be replaced a next info message, is it right? Or should we display both simultaneously?

Ah.. that's an interesting observation. Nice catch. I would want to keep both of them.

**Comment by ahmadabdolsaheb** [comment]
I agree with @raisedadead and @ValeraS That would be optimal.  Should the extra functionality be discussed in another PR?

**Review by raisedadead — APPROVED** [review]
Nice work @hemakshis, thanks for your patience with this PR.

**Comment by raisedadead** [comment]
I have updated the PR with small tweaks that I thought would make it more apparent that new notifications are replacing the older ones, @ahmadabdolsaheb please open a issue to track and implement separate error/info as suggested by Valeriy.

**Comment by raisedadead** [comment]
#### :sparkles: :sparkles: **LGTM.** :sparkles: :sparkles: 

Thanks @hemakshis 

Happy Contributing!

**Pull request merged**